src/linalg.jl

```diff
@@ -208,17 +208,24 @@ end
 
 #--------------------------------------------------
 # Norms
-@inline LinearAlgebra.norm_sqr(v::StaticVector) = mapreduce(abs2, +, v; init=zero(real(eltype(v))))
+function _inner_eltype(v::AbstractArray)
+    return eltype(v) <: AbstractArray ? _inner_eltype(first(v)) : eltype(v)
```

Comment:
Can't you just return `_inner_eltype(first(v))` here? There should be no need to call `eltype`, which might be an abstract type. Note that, for example, this works:
```julia
julia> norm(Vector{Number}[Number[1.0, 2]])
2.23606797749979
```

Reply:
Yep, good point - this would have broken at abstractly typed containers. Your suggestion would work fine for non-empty arrays (of course, it would lead to type-unstability during summation for arrays like e.g. `Any[1, 1.0]` but that seems inevitable).

However, it would break for empty arrays. One way out would be then branch on `isempty(v)` and return `eltype(v)` in that case - but then the whole thing is type-unstable (even for non-abstractly typed iterators). Any ideas?

Reply:
Having a branch for `isempty` case is exactly what `LinearAlgebra` stdlib does: https://github.com/JuliaLang/julia/blob/f8d04b27a7514ecd6e9deb4f605f96984075c773/stdlib/LinearAlgebra/src/generic.jl#L625 . This should be type stable in most cases -- `zero(Any)` is undefined so it doesn't cause problems, and if `eltype(v)` is concrete then it's also fine. The problem is when `eltype(v)` is abstract and has `zero` defined, and `zero(eltype(v))` is of different type than in the non-empty `v` path.